Make use of centralManagerDidUpdateState

fotiDim opened the pull request:
Bluetooth status does not get reported accurately on first run since CBCentralManager needs some warm up time. The `centralManagerDidUpdateState` delegate method which is the one responsible for notifying for Bluetooth status changes was not used before. WIth this PR `isBluetoothAvailable` will make use of that delegate method.

Sunbreak commented on packages/quick_blue/darwin/QuickBlueDarwin.swift:
```diff
@@ -68,11 +68,17 @@ public class QuickBlueDarwin: NSObject, FlutterPlugin {
     case "isBluetoothAvailable":
-      result(manager.state == .poweredOn)
+      print("isBluetoothAvailable: \(manager.state)")  // Needed to trigger CBCentralManager initialization
+      isBluetoothAvailableResult = result
```
What about `messageConnector` for callback from native to Flutter?

Sunbreak commented:
Actually there is an event for `onAvailabilityChanged` in notepad_core_web

https://github.com/woodemi/notepad_core/blob/b0e329f3d6e02f14f8a0e5e48a6ddb48e026b658/notepad_core_web/lib/notepad_core_web.dart#L26-L31

I'd prefer to implementing the same way into mobile & desktop platforms

fotiDim commented:
@Sunbreak I gave your suggestion a try in [another branch](https://github.com/fotiDim/quick.flutter/blob/6720142d99d875ef9220dac6fc184c81dc592722/packages/quick_blue/lib/src/method_channel_quick_blue.dart#L100). I have added a TODO where I see blocker. The message we get needs somehow to be returned from isBluetoothAvailable() in Dart. Is there any way this can be done?

If I understand correctly messageConnector is used to notify from the native side to the Dart side whenever a change happens (sort of like an Event Channel). I think that the two implementations are not mutually exclusive. This PR only fixes isBluetoothAvailable(). We would need another feature-PR that adds `QuickBlue.setAvailabilityHandler()` which notifies the Dart side whenever there is a change in Bluetooth availability (unrelated to isBluetoothAvailable()).

Sunbreak commented:
Linux has a similiar `uninitialized` status to be fixed with `isBluetoothAvailable()`

https://github.com/woodemi/quick.flutter/blob/d918840a0585b4e8a35a460923dc14426e3b84e8/packages/quick_blue/lib/src/quick_blue_linux.dart#L45-L51

fotiDim commented:
@Sunbreak after doing more research:
- If I use messageConnector I would have to add an extra Stream on the Dart side, like [here](https://github.com/woodemi/quick.flutter/pull/27). Since this is an iOS specific solution I believe the "logic" should be in the native code and not in Dart.
- There is no way to do something similar like `ensureInitialized()` for `CBCentralManager` in Swift. We have to use the delegate method.
- If you think that the proposed implementation it too complex an alternative dirty way to solve this would be to add a check when requesting availability and if CBCentralManager is nil initialize and return the availability after a certain delay. Not a big fan of it but it's an option.

I still believe that this PR is the way to go. Let me know what you think.

fotiDim commented:
@Sunbreak I am a bit confused. As I said on iOS there is no straightforward way to have something like `ensureInitialized()`. We have to use the delegate methods and this requires some business logic to be written if you want to have `isBluetoothAvailable` as an async call.

> It is fair for a cross-platform plugin to treat different platforms equally

Yes, that is exactly my intention. I added some Swift logic so we can have a unified API in Dart.

> I'd like to do something like messageConnector in Linux to notify change of availability

Once you add the availability notification for Linux, I can also add it for iOS/MacOS. But this a separate feature and out of scope for this PR. This PR only fixes `isBluetoothAvailable` on iOS. Perhaps the title was misleading so I changed it. If you have any specific suggestion on how else this can be fixed please let me know.

Sunbreak commented:
`isBluetoothAvailable` is to check **_at the time_** whether Bluetooth is available. And `manager.state != .powerOn` is the correct answer

If users of the API would like to wait some time until Bluetooth is available, they should wait by themselves in Dart/Flutter with `Future.delay`

```dart
    Future.delayed(const Duration(seconds: 5), () {
      QuickBlue.isBluetoothAvailable();
    });
```

NOT a magic duration inside Swift implementation

```swift
      DispatchQueue.main.asyncAfter(deadline: .now() + 5) { [weak self] in  // The expected result will come from the centralManagerDidUpdateState delegate method. This is just a timeout in case the delegate never gets called
        self?.isBluetoothAvailableResult?(false)
        self?.isBluetoothAvailableResult = nil
      }
```

Sunbreak commented:
A more elegant way to handle the need of users, is to add listern function for `onAvailabilityChange` event

fotiDim commented:
@Sunbreak I created a [new PR](https://github.com/woodemi/quick.flutter/pull/37). I might close this if the other one is accepted.

Sunbreak commented:
Closing in favor of #37
